tagged/tagged.go

```diff
@@ -0,0 +1,127 @@
+package tagged
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// Value holds an arbitrary value with associated tags
+type Value[T comparable] struct {
+	Value T
+	Tags  []string
+}
+
+// HasTag indicates the Value has a tag matching one or more of the provided arguments
+func (t Value[T]) HasTag(tags ...string) bool {
+	for _, tag := range tags {
+		for _, existing := range t.Tags {
+			if tag == existing {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// New returns a tagged value, that can be added to a Values collection
+func New[T comparable](value T, tags ...string) Value[T] {
+	return Value[T]{
+		Value: value,
+		Tags:  tags,
+	}
+}
+
+// Values is a utility to handle a set of tagged items including basic filtering
+type Values[T comparable] []Value[T]
+
+// Collect returns a slice containing the values in the set
+func (t Values[T]) Collect() []T {
+	out := make([]T, len(t))
+	for i, v := range t {
+		out[i] = v.Value
+	}
+	return out
+}
+
+// HasTag indicates this set contains one or more items matching any of the provided tags
+func (t Values[T]) HasTag(tags ...string) bool {
+	for _, tagged := range t {
+		if tagged.HasTag(tags...) {
+			return true
+		}
+	}
+	return false
+}
+
+// HasValue indicates any of the provided values is present in this set
+func (t Values[T]) HasValue(value ...T) bool {
+	for _, tagged := range t {
+		for _, v := range value {
+			if isEqual(tagged.Value, v) {
```

Comment:
since `T` is `comparable`, `tagged.Value == v` works here, but unfortunately may still result in a panic by the go runtime in certain cases, so I've left a function to compare items if this needs to get worked around, it might be better to just make this use `==`